### PR TITLE
Perf: emulated BN254 pairing

### PR DESCRIPTION
**Pull request by yelhousni**
Taking into account the new range check gadget, this PR removes almost 1M R1CS constraints from the emulated BN254 pairing (verified in a BN254 SNARK). It mostly follows ["Pairings in R1CS" paper](https://eprint.iacr.org/2022/1162.pdf).

### CODE REVIEW

**Comment by yelhousni** [comment]
This branch is based on #532 which needs https://github.com/ConsenSys/gnark-crypto/pull/354 to be merged on master on gnark-crypto side to expose BN254 tower there.

**Comment by ivokub** [comment]
> This branch is based on #532 which needs [ConsenSys/gnark-crypto#354](https://github.com/ConsenSys/gnark-crypto/pull/354) to be merged on master on gnark-crypto side to expose BN254 tower there.

Shouldn't `go get -u github.com/consensys/gnark-crypto@develop` be sufficient?

**Review by ivokub — APPROVED** [review]
Looks perfect! I only made some modifications to keep the interfaces similar, and didn't change anything in the computation itself.

Before merging I would just rebase on top of develop just in case to see we're not reintroducing regressions. There have been several updates to field emulation. If you are OK with it then I can do it myself and force push.

However, when I run TestPairTestSolve locally in test engine then 10% cases fail. I'll have a look to understand if it is due to field emulation. I would postpone merging until then.

**Comment by yelhousni** [comment]
@ivokub rebased on develop which includes now your #579 subtraction bug fix, so PR is ready!

**Review by ivokub — APPROVED** [review]
Very good! I only removed the test which runs the profiler.

File: std/algebra/emulated/sw_bn254/doc.go [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Package sw_bn254 implements G1 and G2 arithmetics and pairing computation over BN254 curve.
 //
-// The implementation follows [[Housni22]]: "Pairings in Rank-1 Constraint Systems".
+// The implementation follows [Housni22]: "Pairings in Rank-1 Constraint Systems".
```
Comment (ivokub) [inline std/algebra/emulated/sw_bn254/doc.go:3]: Yup, I think `[Housni22]` is correct, then I just wanted to have academic style reference for documentation site (example with `[[Housni22]]`. The inner brackets is for link and outer bracket for display):
image

**Pull request merged**